packages/ref-core/src/ref_core/constraints.py

```diff
@@ -0,0 +1,148 @@
+from typing import Protocol, runtime_checkable
+
+import pandas as pd
+from attrs import frozen
+from loguru import logger
+
+from ref_core.exceptions import ConstraintNotSatisfied
+
+
+@runtime_checkable
+class GroupValidator(Protocol):
+    """
+    A constraint that must be satisfied when executing a given metric run.
+
+    All constraints must be satisfied for a given group to be run.
+    """
+
+    def validate(self, group: pd.DataFrame) -> bool:
+        """
+        Validate if the constraint is satisfied by the dataset.
+
+        This is executed after the apply method to determine if the constraint is satisfied.
+        If the constraint is not satisfied, the group will not be executed.
+
+        Parameters
+        ----------
+        group
+            A group of datasets that is being validated.
+
+        Returns
+        -------
+        :
+            Whether the constraint is satisfied
+        """
+        ...
+
+
+@runtime_checkable
+class GroupOperation(Protocol):
+    """
+    An operation to perform on a group of datasets resulting in a new group of datasets.
```

Comment:
> group of datasets

I assume the group of datasets here is not the same as the group above? Otherwise, I'm probably struggling with the language "the group will not be executed" because datasets can't be executed, right?